### DISCUSSION TITLE
Any additional dicom metadata for tio.Image ?

### DISCUSSION THREAD

**Discussion by jamesjjcondon** [TorchIO-project/torchio#439] [Q&A] [answered]
Hi, 

I'm exploring torchio for mammograms. 

Re tio.Image

when the source docs [say](https://github.com/fepegar/torchio/blob/58536b48bc1b04333142df9c1ecdbf83df2559a0/torchio/data/image.py#L69):
"
**kwargs: Items that will be added to the image dictionary, e.g.
            acquisition parameters."

How do I call this?

My dicoms have tags like this:
(0028, 1050) Window Center                       DS: [1802.310000, 1802.310000, 2020.704000, 1583.916000]

Can I reference any dicom metadatum load when I call tio.Image() ?

Thanks

**Comment by fepegar** [comment, marked as answer]
Hi, @jamesjjcondon. That just means that you can add your own keys and values to `Image`. If you want to keep the DICOM tags, we have multiple options:
1. We modify `Image` and `io.read_image` so the DICOM tags are kept. This would mean changing multiple things in the library
2. You create your own [custom image class](https://github.com/fepegar/torchio/issues/437), overriding the `load` method
3. You write a function `tags_dict = get_metadata(path)` that you call before instantiating the image, then you can do e. g. `image = tio.ScalarImage(path, **tags_dict)`

I would go for the 3rd one. I think you can easily get the tags using `pydicom` or even `SimpleITK`.

Would any of these work?

**Reply by jamesjjcondon** [reply]
Thanks, 

3. Is double reading which will be too slow for me. 

Maybe 2. will work. If I get a chance, may submit a PR if you think it's worth it?
TorchIO could be very useful and widespread but currently seems to focus on brain.

Thanks again.

**Reply by fepegar** [reply]
Some formats and libraries support reading only the header, which is very fast. That's why I suggested option number 3. Also, that would be easiest to implement.

Thanks for your offer to contribute! Before submitting a PR, I think we can first discuss the approach you've followed and the changes it would mean to the library.

I agree it probably seems biased towards neuroimaging. The reason is that that's my research field. I'm hoping to make it more open to other modalities.

Out of curiosity, why do you need to access the DICOM tags?

**Reply by jamesjjcondon** [reply]
Yep, realised that without actually reading the image the metadata should be pretty quick to load with pydicom.
I'll leave a PR for now.

I have to make a custom transform which inverts an image based on a WindowCenter and WindowDepth which are mammogram metadata and different for each image.

**Reply by fepegar** [reply]
I see. In that case, maybe you can even read the values within the transform:

```python
import torchio as tio


class WindowTransform(tio.IntensityTransform):
    def apply_transform(self, subject):
        for image in self.get_images(subject):
            center, depth = self.get_center_depth(image.path)
            transformed = self.apply_window(image.data, center, depth)
            image.set_data(transformed)
        return subject
    
    @staticmethod
    def get_center_depth(path):
        # read metadata from path
        return center, depth
    
    @staticmethod
    def apply_window(tensor, center, depth):
        # do your processing
        return tensor

```
